index.js

```diff
@@ -0,0 +1,92 @@
+const url = require('url');
+
+/**
+ * Gate requests based on CORS data. For requests that are not permitted via CORS, invoke the
+ * failure options callback, which defaults to rejecting the request.
+ *
+ * @param {Object} options
+ * @param {String} options.origin The origin of the server - requests from this origin will always
+ *   proceed.
+ * @param {Boolean=} options.strict Whether to reject requests that lack an Origin header. Defaults
+ *   to true.
+ * @param {Boolean=} options.allowSafe Whether to enforce the strict mode for safe requests (HEAD,
+ *   GET). Defaults to true.
+ * @param {Function=} options.failure A standard connect-style callback for handling failure.
+ *   Defaults to rejecting the request with 403 Unauthorized.
+ */
+function corsGate(options) {
+  options = Object.assign({
+    strict: true,
+    allowSafe: true,
+    failure(req, res, next) {
+      // Set `statusCode` vs. using `res#status`, as https://github.com/expressjs/cors does, so this
+      // will work with any Connect-compatible server.
+      res.statusCode = 403;
+      res.end();
+    }
+  }, options);
+
+  if (typeof options.origin !== 'string') {
+    throw new Error("Must specify the server's origin.");
+  }
+
+  const thisOrigin = options.origin.toLowerCase();
+  const failure = options.failure;
+
+  return function(req, res, next) {
+    const origin = (req.headers.origin || '').toLowerCase().trim();
+
+    if (!origin) {
+      // Fail on missing origin when in strict mode, but allow safe requests if allowSafe set.
+      if (options.strict && (!options.allowSafe || ['GET', 'HEAD'].indexOf(req.method) === -1)) {
+        return failure(req, res, next);
+      }
+
+      // The `origin` header will always be sent for browsers that support CORS:
+      // https://developer.mozilla.org/en/docs/HTTP/Access_control_CORS#Origin
+      // If the header is missing, this is not a browser request and thus we shouldn't
```

Comment:
Please take this comment out, I made it before conducting further research and finding [these exceptions](https://docs.google.com/document/d/1DJdcCAlsSHRrd3vlsX1L01WFl_de5LU2ydszXt-G9g8/edit#bookmark=id.e8pzqb95rr6n).